hack/add-version/main.go

```diff
@@ -203,120 +278,120 @@ func newVersionManager(ctx context.Context, region, bucket, distributionID strin
+
 	return err
 }
 
-func (m *versionManager) invalidateCaches(ctx context.Context, version string) error {
-	major := semver.Major(version)
-	minor := semver.MajorMinor(version)
-	invalidation, err := m.cloudfrontc.CreateInvalidation(ctx, &cloudfront.CreateInvalidationInput{
+func (m *versionManager) invalidateCaches(ctx context.Context, ver version) error {
+	invalidIn := &cloudfront.CreateInvalidationInput{
 		DistributionId: aws.String(m.distributionID),
 		InvalidationBatch: &cftypes.InvalidationBatch{
 			CallerReference: aws.String(fmt.Sprintf("%d", time.Now().Unix())),
 			Paths: &cftypes.Paths{
 				Quantity: aws.Int32(2),
 				Items: []string{
-					"/" + versionJSONPath("major", major),
-					"/" + versionJSONPath("minor", minor),
+					"/" + ver.URL(granularityMajor),
+					"/" + ver.URL(granularityMinor),
 				},
 			},
 		},
-	})
+	}
+	invalidation, err := m.cloudfrontc.CreateInvalidation(ctx, invalidIn)
 	if err != nil {
 		return err
 	}
+
 	waiter := cloudfront.NewInvalidationCompletedWaiter(m.cloudfrontc)
-	if err := waiter.Wait(ctx, &cloudfront.GetInvalidationInput{
+	waitIn := &cloudfront.GetInvalidationInput{
 		DistributionId: aws.String(m.distributionID),
 		Id:             invalidation.Invalidation.Id,
-	}, maxCacheInvalidationWaitTime); err != nil {
+	}
+	if err := waiter.Wait(ctx, waitIn, maxCacheInvalidationWaitTime); err != nil {
 		return err
 	}
+
 	return nil
 }
 
-func granularityFromVersion(version string) (string, error) {
-	switch {
-	case semver.Major(version) == version:
-		return "major", nil
-	case semver.MajorMinor(version) == version:
-		return "minor", nil
-	case semver.Canonical(version) == version:
-		return "patch", nil
-	default:
-		return "", fmt.Errorf("invalid version %q", version)
+func waitForCacheUpdate(ctx context.Context, updateFetcher *versionsapi.Fetcher, ver version, log *logger.Logger) {
+	sawAddedVersions := true
+	if err := ensureMinorVersionExists(ctx, updateFetcher, ver); err != nil {
+		sawAddedVersions = false
+		log.Warnf("Failed to ensure minor version exists: %v. This may be resolved by waiting.", err)
+	}
+
+	if err := ensurePatchVersionExists(ctx, updateFetcher, ver); err != nil {
+		sawAddedVersions = false
+		log.Warnf("Failed to ensure patch version exists: %v. This may be resolved by waiting.", err)
+	}
+
+	if sawAddedVersions {
+		log.Infof("Versions are available via API.")
 	}
 }
 
-func versionJSONPath(granularity, base string) string {
-	return path.Join(constants.CDNVersionsPath, "stream", stream, granularity, base, imageKind+".json")
+func listJSONPath(list *versionsapi.List) string {
+	return path.Join(constants.CDNVersionsPath, "stream", list.Stream, list.Granularity, list.Base, imageKind+".json")
 }
 
-func versionURL(granularity, base string) string {
-	return constants.CDNRepositoryURL + "/" + versionJSONPath(granularity, base)
+type granularity int
+
+const (
+	granularityMajor granularity = iota
+	granularityMinor
+	granularityPatch
```

Comment:
```suggestion
```

This is currently unused as we  only map from 

- major -> minor
- minor -> patch

So i would probably want to remove it. But this is really just a  nit.